src/table/body-cell/index.tsx

```diff
@@ -2,72 +2,128 @@
-import React from 'react';
+import React, { useCallback, useRef } from 'react';
+import useFocusVisible from '../../internal/hooks/focus-visible';
+import { useEffectOnUpdate } from '../../internal/hooks/use-effect-on-update';
+import Button from '../../button/internal';
+import { ButtonProps } from '../../button/interfaces';
 import { TableProps } from '../interfaces';
-import { useVisualRefresh } from '../../internal/hooks/use-visual-mode';
+import { TableTdElement, TableTdElementProps } from './td-element';
+import { InlineEditor } from './inline-editor';
+import { useStableScrollPosition } from './use-stable-scroll-position';
 
-interface TableBodyCellProps {
-  className?: string;
-  style?: React.CSSProperties;
-  wrapLines: boolean | undefined;
-  isFirstRow: boolean;
-  isLastRow: boolean;
-  isEvenRow?: boolean;
-  stripedRows?: boolean;
-  isSelected: boolean;
-  isNextSelected: boolean;
-  isPrevSelected: boolean;
-  children?: React.ReactNode;
-  hasSelection?: boolean;
-  hasFooter?: boolean;
+const readonlyState = Object.freeze({
+  isEditing: false,
+  currentValue: undefined,
+  setValue: () => {},
+});
+
+const submitHandlerFallback = () => {
+  throw new Error('The function `handleSubmit` is required for editable columns');
+};
+
+interface TableBodyCellProps<ItemType, ValueType> extends TableTdElementProps {
+  column: TableProps.EditableColumnDefinition<ItemType, ValueType>;
+  item: ItemType;
+  isEditing: boolean;
+  onEditStart: () => void;
+  onEditEnd: () => void;
+  submitEdit?: TableProps.SubmitEditFunction<ItemType, ValueType>;
+  ariaLabels: TableProps['ariaLabels'];
 }
 
-export function TableBodyCell({
+function TableCellEditable<ItemType, ValueType>({
   className,
-  style,
-  children,
-  wrapLines,
-  isFirstRow,
-  isLastRow,
-  isSelected,
-  isNextSelected,
-  isPrevSelected,
-  isEvenRow,
-  stripedRows,
-  hasSelection,
-  hasFooter,
-}: TableBodyCellProps) {
-  const isVisualRefresh = useVisualRefresh();
+  item,
+  column,
+  isEditing,
+  onEditStart,
+  onEditEnd,
+  submitEdit,
+  ariaLabels,
+  isVisualRefresh,
+  ...rest
+}: TableBodyCellProps<ItemType, ValueType>) {
+  const editActivateRef = useRef<ButtonProps.Ref>(null);
+  const cellRef = useRef<HTMLTableCellElement>(null);
+  const focusVisible = useFocusVisible();
+  const { storeScrollPosition, restoreScrollPosition } = useStableScrollPosition(cellRef);
+
+  const handleEditStart = () => {
+    storeScrollPosition();
+    if (!isEditing) {
+      onEditStart();
+    }
+  };
+
+  const scheduleRestoreScrollPosition = useCallback(
+    () => setTimeout(restoreScrollPosition, 0),
+    [restoreScrollPosition]
+  );
+
+  const tdNativeAttributes = {
+    ...(focusVisible as Record<string, string>),
+    onFocus: scheduleRestoreScrollPosition,
+    'data-inline-editing-active': isEditing.toString(),
+  };
+
+  useEffectOnUpdate(() => {
+    if (!isEditing && editActivateRef.current) {
+      editActivateRef.current.focus({ preventScroll: true });
```

Comment:
Why do you prevent native scroll and schedule manual scroll right after?

Reply:
P.S. This entire scroll management story produced at a lot of non-idiomatic code. Let's talk in DM more about reasons for it and how to convert into more idiomatic code